Turbo 8 Morphing Support

adubov1 opened the pull request:
This simply adds listeners for Turbo 8 Morphing in the same way there are already listeners for Frame/Drive rendering. Please let me know if you'd like to add any documentation or tests!

SimoTod commented:
Thanks, we don't need any documentation since it's intended to be a zero configuration script that you just include and forget about it.
We should probably add some tests (also for me to understand the feature, I don't use turbo so I'm not much aware of relatively recent features). I'm actually surprised that the plugin is still used with me having done zero maintenance during the last year. 😅

adubov1 commented:
No problem at all! I needed this to work with the new morphing stuff for my own project, so why not improve upstream. I just found an fixed a bug related to the `before-morph-element` event having a `newElement` property in the detail rather than a `newBody`. I also realized that this callback doesn't need to be performed for attribute morphing, so I removed that listener. 

I will be adding tests soon!

adubov1 commented:
Just added the test! I also had to fix the old tests. For some reason Turbo is not working from the Skypack CDN anymore. Super weird since its what is recommended in the Turbo docs. I switched all of them to jsDelivr. I also noticed the streams test you had skipped passed just fine for me, so I removed the skip. Also, I changed the `data-turbolinks-permanent` attribute to the newer `data-turbo-permanent`.

adubov1 commented:
I also just found and fixed a bug where Alpine controllers were in a partially-initialized state after a Turbo morph. After a morph happened to an `x-data` element, it would be partially broken until it was interacted with. 

For example, an element with an `x-show` would be shown when it shouldn't be, since its `x-cloak` would already have been turned into a `data-alpine-was-cloaked`. I assume other random functionality would also break all over the place.

I fixed this by reinitializing any `x-data` elements that were morphed.

adubov1 commented:
@SimoTod I believe this is good to go. Would you like to see any more tests made for this?

SimoTod commented:
Thanks, I'll try to review as soon as possible this week and let you know. 👍

misaacs21 commented:
Is there anything preventing this from being merged in? I've found I need this functionality as well.

SimoTod commented:
@misaacs21 just me having the memory span of a goldfish. :)
I've just had a look, I don't love the destory+init tree although I see what @adubov1 mean. 
I'll play with it a little bit but I'll try to get it merged soon. Thanks for reminding me.

SimoTod commented on tests/integration/turbo/morph.cy.js:
```diff
@@ -0,0 +1,45 @@
+      body: refreshResponse
+    }).as('Refresh')
+
+    cy.visit('/tests/res/turbo/stream/index.html')
```
Apologise, I know it took longer then expected but I can't follow this test. why is it loading files from the steam folder? I was expecting a simple test like the one in the cloak folder (maybe just slightly more complex) with the morph swapping strategy. This test seems to do deal with streaming instead

SimoTod replied:
ok, I read a bit more about morph and I can see that it only happens when Turbo detect that the page was the same. However this test initially loads a page without "<meta name="turbo-refresh-method" content="morph">" so is it actually morphing the page or doing a regular swap?

adubov1 replied:
Oops, I accidentally used the template in the stream folder instead of using the one I made. Also, good point, I think one way we can guarantee a morph is if the only change on the page is an attribute on an existing element. I will change the test to do that!

SimoTod commented on src/bridge.js:
```diff
@@ -89,5 +110,8 @@ export default class Bridge {
     document.addEventListener('turbo:before-cache', beforeCacheCallback)
+
+    document.addEventListener('turbo:morph', morphCallback)
+    document.addEventListener('turbo:before-morph-element', beforeMorphCallback)
```
is this callback invoked for each morphed element? I suspect it should just use Alpine.cloneNode instead of destroying the tree?

adubov1 replied:
I will look into this

bjufre commented:
@adubov1 @SimoTod any news here? 🙏 anything I xan do to help?

adubov1 commented:
@bjufre @SimoTod my apologies for the delay, have been very busy. I'll try to get to this soon!

adubov1 commented:
@SimoTod 

So I spent a lot of time diving into Alpine's codebase. I wanted to see if I could use `cloneNode` the same way they do. Unfortunately, no matter what I tried, I could only _either_ get Alpine elements to initialize properly **or** get them to be interactive after a morph. I can get into what I've tried and how if you'd like.

I did make some great improvements to morphing compatibility, though. First of all, I noticed that Alpine's morph plugin and Turbo/Idiomorph have the same issue where morphed shorthand `x-on` (like `@click`) cannot be set by default since `Element.setAttribute` doesn't allow the "@" symbol. I copied over the workaround that Alpine is doing and cited their codebase.

Second, I noticed that Turbo's render events fire for morphs. There is a lot less work we need to do for morphs since we don't need to worry about cache renders. So, I ignored morph renders for the render callbacks. 

I found a workaround to destroying the tree as well. I don't love this approach since it relies on what seems to be private API, but removing the `_x_marker` property from morphed elements tricks Alpine into initializing the tree. There's also a simple check to only reinitialize elements with an Alpine data stack. 

Additionally, I'm running `destroyTree` on data stacks that get removed while morphing. Please let me know if this is unnecessary since they are removed from the dom anyway. I'm just not sure enough about how Alpine handles cleanup, so I wanted to be safe.

I made the tests actually test morphing now 🤦. There are three tests, one for Alpine initializing correctly, one for immediate interactivity (which is the issue one would notice immediately when trying to use this library with morphing), and another for frames since they render/morph slightly differently from regular elements.

SimoTod commented:
Thanks for your time. I think this version is clean enough and should do what you expect. If you just move those files from cypress/fixtures/morph into examples/morph, I'm happy to merge it and tag a release

adubov1 commented on cypress/fixtures/morph/initialization.html:
```diff
@@ -0,0 +1,31 @@
+<html>
```
I use those in the tests. I put them separately since they are entire pages of HTML, so they were making the tests a bit less readable. I can put them in the test if you'd prefer, though

SimoTod replied:
Fair enough, I got a bit confused because they look the same as the files in the res folder. I'll probably just move them and update the config but don't worry about this.
I'll try to merge and tag tonight (I'm replying from my phone). Thanks for this

SimoTod commented:
Tagged and published to npm (https://www.npmjs.com/package/alpine-turbo-drive-adapter)

Thanks
